proposer/fp/src/main.rs

```diff
@@ -0,0 +1,505 @@
+
+        Ok(OutputRoot {
+            root,
+            l2_block_number,
+        })
+    }
+
+    async fn poll_safe_l2_head(&self) -> Result<U256> {
+        let response: serde_json::Value = self
+            .l2_client
+            .request(
+                "eth_getBlockByNumber",
+                vec![serde_json::json!("safe"), serde_json::json!(false)],
+            )
+            .await?;
+
+        let block_number = response
+            .get("number")
+            .and_then(|v| v.as_str())
+            .ok_or_else(|| eyre::eyre!("Invalid block response format"))?;
+
+        // Remove "0x" prefix and parse as hex
+        let block_number = U256::from_str(&block_number.replace("0x", ""))?;
+        Ok(block_number)
+    }
+
+    async fn get_output_root_at_block(&self, block_number: U256) -> Result<H256> {
+        let params = vec![serde_json::json!(format!("0x{:x}", block_number))];
+
+        let response: serde_json::Value = self
+            .l2_node_client
+            .request("optimism_outputAtBlock", params)
+            .await?;
+
+        tracing::debug!("RPC Response: {:?}", response);
+
+        // The outputRoot is directly in the response, not under "result"
+        let output_root = response
+            .get("outputRoot")
+            .and_then(|v| v.as_str())
+            .ok_or_else(|| eyre::eyre!("Invalid output root response format: {:?}", response))?;
+
+        tracing::debug!("Extracted output root: {}", output_root);
+
+        // Remove "0x" prefix and parse as H256
+        let output_root = output_root.strip_prefix("0x").unwrap_or(output_root);
+        Ok(H256::from_str(output_root)?)
+    }
+
+    async fn create_new_game(&self, l2_block_number: U256) -> Result<()> {
+        let function = Function {
+            name: "create".to_string(),
+            inputs: vec![
+                Param {
+                    name: "_gameType".to_string(),
+                    kind: ParamType::Uint(32),
+                    internal_type: None,
+                },
+                Param {
+                    name: "_rootClaim".to_string(),
+                    kind: ParamType::FixedBytes(32),
+                    internal_type: None,
+                },
+                Param {
+                    name: "_extraData".to_string(),
+                    kind: ParamType::Bytes,
+                    internal_type: None,
+                },
+            ],
+            outputs: vec![],
+            constant: None,
+            state_mutability: StateMutability::Payable,
+        };
+
+        let root_claim = self.get_output_root_at_block(l2_block_number).await?;
+        let extra_data = ethers::abi::encode(&[Token::Uint(l2_block_number)]);
+
+        let params = vec![
+            Token::Uint(GAME_TYPE.into()),
+            Token::FixedBytes(root_claim.as_bytes().to_vec()),
+            Token::Bytes(extra_data),
+        ];
+        let data = function.encode_input(&params)?;
```

Comment:
We can use the `sol!` macro from `alloy`, and contract bindings to simplify this logic.